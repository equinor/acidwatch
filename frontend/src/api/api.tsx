import config from "../configuration";
import { SimulationResults } from "../dto/SimulationResults";
import { Project } from "../dto/Project";
import { Simulation } from "../dto/Simulation";
import { ModelConfig } from "../dto/FormConfig";
import { ExperimentResult } from "../dto/ExperimentResult";
import { getUserToken } from "../services/auth";
import { getAccessToken } from "../services/auth";

export type concentrations = {
    [key: string]: number;
};

export const runSimulation = async (
    concentrations: Record<string, number>,
    parameters: Record<string, number>,
    modelId: string
): Promise<SimulationResults> => {
    const apiUrl = `${config.API_URL}/models/${modelId}/runs`;
    const token = await getAccessToken();

    // Set up timeout
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 500 * 1000);

    try {
        const response = await fetch(apiUrl, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${token}`,
            },
            body: JSON.stringify({
                concs: concentrations,
                settings: parameters,
            }),
            signal: controller.signal,
        });

        clearTimeout(timeout);
        if (!response.ok) {
            throw new Error("Network error");
        }

        return response.json();
    } catch (error) {
        if ((error as Error).name === "AbortError") {
            throw new Error("Request timed out");
        }
        throw error;
    }
};

export const getModels = async (): Promise<ModelConfig[]> => {
    const token = await getAccessToken();
    const response = await fetch(config.API_URL + "/models", {
        method: "GET",
        headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
        },
    });

    if (!response.ok) {
        throw new Error("Network error");
    }

    const data = await response.json();

    return data;
};

export const saveProject = async (name: string, description: string, isPrivate: boolean): Promise<string> => {
    const token = await getAccessToken();
    const response = await fetch(config.API_URL + "/project", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + token,
        },
        body: JSON.stringify({
            name,
            description,
            private: isPrivate,
        }),
    });
    if (!response.ok) {
        throw new Error("Network response was not ok");
    }

    const data = await response.json();
    return data.id;
};

export const getProjects = async (): Promise<Project[]> => {
    const token = await getAccessToken();
    const response = await fetch(config.API_URL + "/projects", {
        method: "GET",
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + token,
        },
    });

    if (!response.ok) {
        throw new Error("Network response was not ok");
    }
    const data: Project[] = await response.json();
    return data;
};

export const deleteProject = async (projectId: string) => {
    const token = await getAccessToken();
    console.log("Deleting project with id:", projectId);
    try {
        const response = await fetch(config.API_URL + "/project/" + projectId, {
            method: "DELETE",
            headers: {
                "Content-Type": "application/json",
                Authorization: "Bearer " + token,
            },
        });

        if (!response.ok) {
            throw new Error("Network response was not ok");
        }
    } catch (error) {
        console.error("Error deleting project:", error);
    }
};

export const getSimulations = async (projectId: string): Promise<Simulation[]> => {
    const token = await getAccessToken();
    const response = await fetch(config.API_URL + "/project/" + projectId + "/scenarios", {
        method: "GET",
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + token,
        },
    });

    if (!response.ok) {
        throw new Error("Network response was not ok");
    }
    const data: Simulation[] = await response.json();
    return data;
};
export interface SimulationInput {
    key: string;
    defaultvalue: number;
}

export const saveSimulation = async (
    projectId: string,
    result: SimulationResults | undefined,
    selectedModel: string,
    parameters: Record<string, number> | undefined,
    simulationName: string
): Promise<any> => {
    const token = await getAccessToken();

    const body = JSON.stringify({
        name: simulationName,
        model: selectedModel,
        scenario_inputs: {
            initialConcentrations: result?.initialConcentrations ?? {},
            parameters: parameters,
        },
    });
    const response = await fetch(config.API_URL + "/project/" + projectId + "/scenario", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + token,
        },
        body: body,
    });
    if (!response.ok) {
        throw new Error("Network response was not ok");
    }

    const data = await response.json();
    return data;
};
export const deleteSimulation = async (projectId: string, simulationId: number): Promise<void> => {
    const token = await getAccessToken();
    const response = await fetch(`${config.API_URL}/project/${projectId}/scenario/${simulationId}`, {
        method: "DELETE",
        headers: {
            Authorization: "Bearer " + token,
        },
    });

    if (!response.ok) {
        throw new Error("Network response was not ok");
    }
};

export const saveResult = async (
    projectId: string,
    results: SimulationResults,
    simulationId: string
): Promise<void> => {
    const body = JSON.stringify(results);

    const token = await getAccessToken();
    const response = await fetch(`${config.API_URL}/project/${projectId}/scenario/${simulationId}/result`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + token,
        },
        body: body,
    });

    if (!response.ok) {
        throw new Error("Network response was not ok");
    }
};

export const getSimulationResults = async (projectId: string, simulationId: string): Promise<SimulationResults> => {
    const token = await getAccessToken();
    const response = await fetch(`${config.API_URL}/project/${projectId}/scenario/${simulationId}/results`, {
        method: "GET",
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + token,
        },
    });

    if (!response.ok) {
        throw new Error("Network response was not ok");
    }

    const data = await response.json();

    const simulationResults: SimulationResults = data[0];

    return simulationResults;
};

export async function switchPublicity(projectId: string): Promise<any> {
    const url = `${config.API_URL}/project/${projectId}/switch_publicity`;
    const token = await getAccessToken();
    console.log("Switching project publicity:", projectId);
    try {
        const response = await fetch(url, {
            method: "PUT",
            headers: {
                "Content-Type": "application/json",
                Authorization: "Bearer " + token,
            },
        });

        if (!response.ok) {
            throw new Error(`Error updating project: ${response.statusText}`);
        }

        const data = await response.json();
        return data;
    } catch (error) {
        console.error("Error updating project:", error);
        throw error;
    }
}

export const extractAndReplaceKeys = (pattern: string, replacement: string, dictionary: Record<string, any>) => {
    return Object.keys(dictionary)
        .filter((key) => key.startsWith(pattern))
        .reduce<Record<string, number>>((acc, key) => {
            acc[key.replace(pattern, replacement)] = dictionary[key];
            return acc;
        }, {});
};

const processData = (response: any): ExperimentResult[] => {
    const experimentResults: ExperimentResult[] = response.flatMap((item: any) => {
        const experimentResult = item.data.labData.concentrations.entries.map((entry: any) => {
            const species = entry.species;

            const inputConcentrations = extractAndReplaceKeys("In_", "", species);
            const inputConcentrationsCapitalized = Object.fromEntries(
                Object.entries(inputConcentrations).map(([key, value]) => [key.toUpperCase(), value])
            );
            const outputConcentrations = extractAndReplaceKeys("Out_", "", species);
            const outputConcentrationsCapitalized = Object.fromEntries(
                Object.entries(outputConcentrations).map(([key, value]) => [key.toUpperCase(), value])
            );
            const experimentResult: ExperimentResult = {
                name: item.data.general.name + "-" + entry.step,
                initialConcentrations: inputConcentrationsCapitalized,
                finalConcentrations: outputConcentrationsCapitalized,
                pressure: item.data.general.pressure ?? null,
                temperature: item.data.general.temperature ?? null,
                time: entry.time ?? null,
            };
            return experimentResult;
        });

        return experimentResult;
    });

    return experimentResults;
};
export async function getLabResults(): Promise<ExperimentResult[]> {
    const token = await getUserToken(config.OASIS_SCOPE);
    const response = await fetch("/oasis/CO2LabResults", {
        method: "GET",
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + token,
        },
    });

    if (!response.ok) {
        if (response.status === 401) {
            throw new Error("Unauthorized: Apply for access to CO2 lab results in AccessIT");
        } else if (response.status === 403) {
            throw new Error(
                "You do not have permission to access this resource. Apply for access to CO2 lab results in AccessIT"
            );
        } else {
            throw new Error("Network response was not ok");
        }
    }
    const data = await response.json();

    const transformedData = processData(data);
    return transformedData;
}
