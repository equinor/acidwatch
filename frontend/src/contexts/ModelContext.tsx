import { createContext, ReactNode, useContext, useEffect } from "react";
import { ModelConfig } from "../dto/FormConfig";
import { getModels } from "../api/api";
import { useQuery, useQueryClient } from "@tanstack/react-query";

import { DefaultError } from "@tanstack/query-core";
import { useIsAuthenticated } from "@azure/msal-react";

type AvailableModelsContextType = {
    models: ModelConfig[];
    error: DefaultError | null;
    isLoading: boolean;
};

const AvailableModelsContext = createContext<AvailableModelsContextType>(null as any);

export const AvailableModelsProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
    const isAuthenticated = useIsAuthenticated();
    const queryClient = useQueryClient();
    const {
        data: models,
        error,
        isLoading,
    } = useQuery({
        queryKey: ["models"],
        queryFn: getModels,
        retry: false,
    });

    useEffect(() => {
        queryClient.invalidateQueries({ queryKey: ["models"] });
    }, [isAuthenticated, queryClient]);

    return (
        <AvailableModelsContext.Provider value={{ models: models ?? [], error, isLoading }}>
            {children}
        </AvailableModelsContext.Provider>
    );
};

export const useAvailableModels = () => useContext(AvailableModelsContext);
