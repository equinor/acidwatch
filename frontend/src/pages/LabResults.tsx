import React, { useState, useMemo } from "react";
import { useQuery, useQueries } from "@tanstack/react-query";
import { EdsDataGrid, Row } from "@equinor/eds-data-grid-react";
import { getLabResults } from "../api/api.tsx";
import ResultScatterGraph from "../components/ResultScatterPlot.tsx";
import { syntheticResults } from "../assets/syntheticResults.tsx";
import { useAvailableModels } from "../contexts/ModelContext.tsx";
import { ExperimentResult_to_ScatterGraphData } from "../functions/Formatting.tsx";
import { Autocomplete, AutocompleteChanges, Button, Card, EdsProvider, Typography } from "@equinor/eds-core-react";
import { runSimulation } from "../api/api";
import { ScatterGraphData } from "../dto/ScatterGraphInput.tsx";

const LabResults: React.FC = () => {
    const initialPrefix = "in-";
    const finalPrefix = "out-";
    const [plotComponents, setPlotComponents] = useState<string[]>([]);
    const [simulationCache, setSimulationCache] = useState<Record<string, ScatterGraphData[]>>({});
    const [selectedExperiments, setSelectedExperiments] = useState<string[]>([]);
    const { models } = useAvailableModels();

    const {
        data: labResults = syntheticResults,
        error,
        isLoading,
    } = useQuery({
        queryKey: ["results"],
        queryFn: () => getLabResults(),
        retry: false,
    });

    const selectedExperimentData = useMemo(
        () => labResults.filter((result) => selectedExperiments.includes(result.name)),
        [labResults, selectedExperiments]
    );

    const simulationQueries = useQueries({
        queries: selectedExperimentData.flatMap((experiment) => {
            const filteredConcs = Object.fromEntries(
                Object.entries(experiment.initialConcentrations).filter(([, value]) => Number(value) !== 0)
            );
            return models
                .filter((model) => model.category === "Primary")
                .filter((model) => Object.entries(filteredConcs).every(([key]) => model.validSubstances.includes(key)))
                .map((model) => ({
                    queryKey: ["simulation", experiment.name, model.modelId],
                    queryFn: async (): Promise<ScatterGraphData[]> => {
                        const cacheKey = `${experiment.name}_${model.modelId}`;
                        if (simulationCache[cacheKey]) {
                            return simulationCache[cacheKey];
                        }
                        try {
                            const simulation = await runSimulation(
                                filteredConcs,
                                model.parameters && Object.keys(model.parameters).length > 0
                                    ? {
                                          pressure: experiment.pressure ?? 0,
                                          temperature: 273 + (experiment.temperature ?? 0), // Convert to Kelvin
                                      }
                                    : {},
                                model.modelId
                            );

                            const result = Object.entries(simulation.finalConcentrations).map(([name, value]) => ({
                                x: name,
                                y: value,
                                label: `${model.displayName} (${experiment.name})`,
                                experimentName: experiment.name,
                                modelName: model.displayName,
                            }));

                            // Save to cache
                            setSimulationCache((prev) => ({
                                ...prev,
                                [cacheKey]: result,
                            }));

                            return result;
                        } catch (error) {
                            console.error(
                                `Simulation failed for ${experiment.name} with model ${model.displayName}:`,
                                error
                            );
                            return [];
                        }
                    },
                    enabled: selectedExperiments.length > 0 && models.length > 0,
                    retry: 1,
                    staleTime: Infinity,
                }));
        }),
        combine: (results) => {
            const allSimResults = results
                .filter((result) => result.isSuccess && result.data)
                .flatMap((result) => result.data);

            const isLoading = results.some((result) => result.isLoading);
            const hasErrors = results.some((result) => result.isError);

            return {
                data: allSimResults,
                isLoading,
                hasErrors,
                errorCount: results.filter((result) => result.isError).length,
                totalQueries: results.length,
            };
        },
    });

    const combinedGraphData = useMemo(
        () => [...ExperimentResult_to_ScatterGraphData(selectedExperimentData), ...simulationQueries.data],
        [selectedExperimentData, simulationQueries.data]
    );

    if (isLoading) return <>Fetching results ...</>;

    let issueRetrievingDataInfo = null;
    if (error) {
        issueRetrievingDataInfo = (
            <Card variant="warning" style={{ margin: "2rem 0" }}>
                <Card.Header>
                    <Card.HeaderTitle>
                        <Typography variant="h5">Error fetching data from Oasis</Typography>
                    </Card.HeaderTitle>
                </Card.Header>
                <Card.Content>
                    <Typography variant="body_short_bold">{error.message}</Typography>
                    <Typography variant="body_short">Using synthetic demo data</Typography>
                </Card.Content>
            </Card>
        );
    }

    const simulationStatusInfo = simulationQueries.totalQueries > 0 && (
        <Card style={{ margin: "2rem 0" }}>
            <Card.Content>
                <Typography variant="body_short">
                    {simulationQueries.isLoading ? "Running simulations..." : ""}
                </Typography>
            </Card.Content>
        </Card>
    );

    const initialConcHeaders = Array.from(
        new Set(labResults.flatMap((entry) => [...Object.keys(entry.initialConcentrations)]))
    );
    const finalConcHeaders = Array.from(
        new Set(labResults.flatMap((entry) => [...Object.keys(entry.finalConcentrations)]))
    );

    const columns = [
        {
            columns: [
                {
                    header: "Experiment",
                    id: "name",
                    accessorKey: "name",
                    size: 200,
                },
                {
                    header: "Time",
                    id: "time",
                    accessorKey: "time",
                    size: 65,
                },
                {
                    header: "Temperature (Â°C)",
                    id: "temperature",
                    accessorKey: "temperature",
                    size: 475,
                },
                {
                    header: "Pressure",
                    id: "pressure",
                    accessorKey: "pressure",
                    size: 65,
                },
            ],
            header: "Meta data",
        },
        {
            columns: initialConcHeaders.map((header) => ({
                accessorKey: initialPrefix + header,
                header: header,
                id: initialPrefix + header,
                size: 65,
            })),
            header: "Input Concentrations",
        },
        {
            columns: finalConcHeaders.map((header) => ({
                accessorKey: finalPrefix + header,
                header: header,
                id: finalPrefix + header,
                size: 65,
            })),
            header: "Output Concentrations",
        },
    ];

    const rows = labResults!.map((entry) => ({
        id: entry.name,
        name: entry.name,
        time: String(entry.time),
        temperature: entry.temperature,
        pressure: entry.pressure,
        ...Object.fromEntries(
            Object.entries(entry.initialConcentrations).map(([key, value]) => [
                initialPrefix + key,
                +Number(value).toPrecision(3),
            ])
        ),
        ...Object.fromEntries(
            Object.entries(entry.finalConcentrations).map(([key, value]) => [
                finalPrefix + key,
                +Number(value).toPrecision(3),
            ])
        ),
        meta: {},
    }));

    const handleRowClick = (
        row: Row<{
            meta: object;
            id: string;
            name: string;
            time: string;
            temperature: number | null;
            pressure: number | null;
        }>
    ) => {
        setSelectedExperiments((prevSelectedExperiments) => {
            return prevSelectedExperiments.includes(row.original.id)
                ? prevSelectedExperiments.filter((key) => key !== row.original.id)
                : [...prevSelectedExperiments, row.original.id];
        });
    };

    const handlePlotComponentsChange = (changes: AutocompleteChanges<string>) => {
        setPlotComponents(changes.selectedItems);
    };

    return (
        <>
            <Typography variant="h1">Lab results</Typography>
            {issueRetrievingDataInfo}
            {simulationStatusInfo}

            <>
                <Typography variant="h2">Plot summary</Typography>
                <ResultScatterGraph
                    graphData={combinedGraphData.filter((d): d is ScatterGraphData => d !== undefined)}
                />

                <Typography variant="h2">Plot per component</Typography>
                <div style={{ width: "500px" }}>
                    <Autocomplete
                        label={"Select multiple components"}
                        options={finalConcHeaders}
                        multiple
                        onOptionsChange={handlePlotComponentsChange}
                    />
                </div>
                <ResultScatterGraph
                    graphData={ExperimentResult_to_ScatterGraphData(selectedExperimentData, plotComponents)}
                />
            </>

            <Typography variant="body_short">Select rows to compare.</Typography>
            <EdsProvider density="compact">
                <EdsDataGrid
                    columns={columns}
                    rows={rows}
                    enableColumnFiltering={true}
                    enableMultiRowSelection
                    enableRowSelection
                    onRowClick={handleRowClick}
                    rowSelectionState={{}}
                    rowStyle={(row) => (row.id in selectedExperiments ? { backgroundColor: "lightblue" } : {})}
                />
            </EdsProvider>
            <Button onClick={() => setSelectedExperiments([])}>Clear</Button>
        </>
    );
};

export default LabResults;
