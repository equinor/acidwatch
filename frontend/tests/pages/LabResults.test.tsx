import { describe, it, expect, vi, beforeEach } from "vitest";
import LabResults from "../../src/pages/LabResults";
import { render, screen, waitFor } from "@testing-library/react";
import { AvailableModelsProvider } from "../../src/contexts/ModelContext.tsx";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { getLabResults, getModels, runSimulation } from "../../src/api/api";

const mockLabResults = [
    {
        name: "Experiment 1",
        time: 100,
        temperature: 25,
        pressure: 1,
        initialConcentrations: { CO2: 0.5, H2O: 0.3 },
        finalConcentrations: { CO2: 0.4, H2O: 0.4 },
    },
    {
        name: "Experiment 2", 
        time: 200,
        temperature: 30,
        pressure: 1.2,
        initialConcentrations: { CO2: 0.6, H2O: 0.2 },
        finalConcentrations: { CO2: 0.3, H2O: 0.5 },
    },
];

const mockGetModels = [
    {
        modelId: "DummyModel",
        displayName: "Dummy Model",
        validSubstances: ["CO2", "H2O"],
        parameters: {"Temperature":{ default: 300 }, "Pressure":{ default: 300 }},
        description: "A dummy model for testing purposes",
        category: "Primary"
    },
];

const mockRunSimulationResult = {
    initialConcentrations: { CO2: 0.5, H2O: 0.3 },
    finalConcentrations: { CO2: 0.4, H2O: 0.4 },
    panels: [],
};

// Mock API functions
vi.mock("../../src/api/api", () => ({
    getLabResults: vi.fn(),
    getModels: vi.fn(),
    runSimulation: vi.fn(),
})); 

// Mock EDS components to avoid DOM API issues
vi.mock("@equinor/eds-core-react", () => ({
    Typography: ({ children, variant, ...props }: any) => (
        <div data-testid={`typography-${variant}`} {...props}>{children}</div>
    ),
    Button: ({ children, onClick, ...props }: any) => (
        <button onClick={onClick} {...props}>{children}</button>
    ),
    Card: ({ children, variant, style, ...props }: any) => (
        <div data-testid={`card-${variant || 'default'}`} style={style} {...props}>{children}</div>
    ),
    Autocomplete: ({ label, options, multiple, ...props }: any) => (
        <div data-testid="autocomplete">
            <label>{label}</label>
            <select multiple={multiple} {...props}>
                {options?.map((option: string) => (
                    <option key={option} value={option}>{option}</option>
                ))}
            </select>
        </div>
    ),
    EdsProvider: ({ children }: any) => <div data-testid="eds-provider">{children}</div>,
    AutocompleteChanges: {} as any,
}));

// Mock EDS DataGrid
vi.mock("@equinor/eds-data-grid-react", () => ({
    EdsDataGrid: ({ columns, rows, onRowClick, ...props }: any) => (
        <div data-testid="eds-data-grid" {...props}>
            <table>
                <thead>
                    <tr>
                        {columns?.map((colGroup: any) => (
                            <th key={colGroup.header}>{colGroup.header}</th>
                        ))}
                    </tr>
                </thead>
                <tbody>
                    {rows?.map((row: any) => (
                        <tr key={row.id} onClick={() => onRowClick?.({ original: row })}>
                            <td>{row.name}</td>
                        </tr>
                    ))}
                </tbody>
            </table>
        </div>
    ),
}));

// Mock the plotting components
vi.mock("../../src/components/ResultScatterPlot.tsx", () => ({
    default: ({ graphData }: { graphData: any[] }) => (
        <div data-testid="scatter-plot">
            Scatter Plot with {graphData?.length || 0} data points
        </div>
    ),
}));

// Mock formatting functions
vi.mock("../../src/functions/Formatting.tsx", () => ({
    ExperimentResult_to_ScatterGraphData: vi.fn(() => [
        { x: "CO2", y: 0.4, label: "Experiment 1" },
        { x: "H2O", y: 0.4, label: "Experiment 1" }
    ]),
}));

// Mock synthetic results
vi.mock("../../src/assets/syntheticResults.tsx", () => ({
    syntheticResults: [
        {
            name: "Synthetic Experiment 1",
            time: 50,
            temperature: 20,
            pressure: 1,
            initialConcentrations: { CO2: 0.3, H2O: 0.2 },
            finalConcentrations: { CO2: 0.2, H2O: 0.3 },
        }
    ]
}));

describe("LabResults Page", () => {
    let queryClient: QueryClient;

    beforeEach(() => {
        // Create a fresh QueryClient for each test
        queryClient = new QueryClient({
            defaultOptions: {
                queries: {
                    retry: false,
                    staleTime: 0,
                },
            },
        });

        // Clear all mocks and set up default implementations
        vi.clearAllMocks();
        vi.mocked(getLabResults).mockResolvedValue(mockLabResults);
        vi.mocked(getModels).mockResolvedValue(mockGetModels);
        vi.mocked(runSimulation).mockResolvedValue(mockRunSimulationResult);
    });
    
    const renderComponent = () => {
        return render(
            <QueryClientProvider client={queryClient}>
                <AvailableModelsProvider>
                    <LabResults />
                </AvailableModelsProvider>
            </QueryClientProvider>
        );
    };
    
    it("renders without crashing", () => {
        const { baseElement } = renderComponent();
        expect(baseElement).toBeDefined();
    });

    it("matches snapshot", () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders all main components when data is loaded", async () => {
        renderComponent();

        await waitFor(
            () => {
                const h1Element = screen.getAllByTestId("typography-h1");
                expect(h1Element.some(el => el.textContent === "Lab results")).toBe(true);
            },
        );
        
        await waitFor(() => {
            const h2Elements = screen.getAllByTestId("typography-h2");
            expect(h2Elements.some(el => el.textContent === "Plot summary")).toBe(true);
            expect(h2Elements.some(el => el.textContent === "Plot per component")).toBe(true);
        });
       
    });
});

/*
 
        // Check for scatter plots
        const scatterPlots = screen.getAllByTestId("scatter-plot");
        expect(scatterPlots.length).toBeGreaterThanOrEqual(2); // Should have at least 2 plots
        
        // Check for autocomplete
        expect(screen.getByTestId("autocomplete")).toBeInTheDocument();
        expect(screen.getByLabelText("Select multiple components")).toBeInTheDocument();
        
        // Check for data grid
        expect(screen.getByTestId("eds-data-grid")).toBeInTheDocument();
        
        // Check for Clear button
        expect(screen.getByRole("button", { name: "Clear" })).toBeInTheDocument();
        
        // Check for table headers in the mocked data grid
        expect(screen.getByText("Meta data")).toBeInTheDocument();
        expect(screen.getByText("Input Concentrations")).toBeInTheDocument();
        expect(screen.getByText("Output Concentrations")).toBeInTheDocument();
        
        // Check for experiment data
        expect(screen.getByText("Experiment 1")).toBeInTheDocument();
        expect(screen.getByText("Experiment 2")).toBeInTheDocument();
        
        // Check for instruction text
        expect(screen.getByText("Select rows to compare.")).toBeInTheDocument();

        */
