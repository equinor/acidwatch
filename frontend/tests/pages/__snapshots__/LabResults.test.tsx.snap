// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LabResults Page > matches snapshot 1`] = `
<DocumentFragment>
  Fetching results ...
</DocumentFragment>
`;

exports[`LabResults Page > renders all main components when data is loaded 1`] = `
<DocumentFragment>
  Fetching results ...
</DocumentFragment>
`;
